src/worker/ingestion/action-matcher.ts

```diff
@@ -0,0 +1,498 @@
+            for (let depthDiff = 1; baseElementIndex - depthDiff >= 0; depthDiff++) {
+                // Subtracting depthDiff as elements are reversed, meaning outer elements have higher indexes
+                const currentElementIndex = baseElementIndex - depthDiff
+                if (
+                    this.checkElementAgainstSelectorPartRequirements(elements[currentElementIndex], part.requirements)
+                ) {
+                    baseElementIndex = currentElementIndex
+                    wasPartMatched = true
+                    break
+                } else if (part.directDescendant) {
+                    break
+                }
+            }
+            if (wasPartMatched) {
+                partIndex++
+            } else {
+                if (part.directDescendant) {
+                    // For a direct descendant we need to check the parent parent again
+                    partIndex--
+                } else {
+                    // Otherwise we just move the base down
+                    baseElementIndex--
+                }
+            }
+        }
+        return wasPartMatched
+    }
+
+    private checkElementAgainstSelectorPartRequirements(element: Element, requirements: Partial<Element>): boolean {
+        if (requirements['text'] && element.text !== requirements.text) {
+            return false
+        }
+        if (requirements['tag_name'] && element.tag_name !== requirements.tag_name) {
+            return false
+        }
+        if (requirements['href'] && element.href !== requirements.href) {
+            return false
+        }
+        if (requirements['attr_id'] && element.attr_id !== requirements.attr_id) {
+            return false
+        }
+        if (requirements['attr_class']) {
+            if (
+                !element.attr_class?.length ||
+                !requirements.attr_class.every((className) => element.attr_class!.includes(className))
+            ) {
+                return false
+            }
+        }
+        if (requirements['nth_child'] && element.nth_child !== requirements.nth_child) {
+            return false
+        }
+        if (requirements['nth_of_type'] && element.nth_of_type !== requirements.nth_of_type) {
+            return false
+        }
+        if (requirements['attributes']) {
+            const { attributes } = element
+            if (!attributes) {
+                return false
+            }
+            for (const [key, value] of Object.entries(requirements.attributes)) {
+                if (attributes[key] !== value) {
+                    return false
+                }
+            }
+        }
+        return true
+    }
+}
+
+class SelectorPart {
+    directDescendant: boolean
+    uniqueOrder: number
+    requirements: Partial<Element>
+
+    constructor(tag: string, directDescendant: boolean, escapeSlashes: boolean) {
+        const SELECTOR_ATTRIBUTE_REGEX = /([a-zA-Z]*)\[(.*)=[\'|\"](.*)[\'|\"]\]/
```

Comment:
A comment here would be useful

Reply:
What should this comment convey?

Reply:
I think with longer regexes an example of what it should match is always useful